mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp

```diff
@@ -511,10 +511,48 @@ propagateVectorizationInfo(TransformMapAttr map, const VectorizationData &input,
     // isn't equal to the accumulated value times the smallest coefficient seen
     // on a dimension being vectorized.
     case TransformType::Embed: {
-      // Since embed coefficients can go in any order, we need them sorted
+      // Since embed coefficients can go in any order, we need them sorted.
+      // We first sort by the Embed coefficient
+      // We tiebreak dimensions with equal coefficients as follows:
+      //  - Dimensions that have vectorization info are processed before those
+      //    that don't.
+      //   - Two vectorization dimensions have ties broken by whether their
+      //      needed coefficient is equal to their embed coefficient.
+      //   - If both inputs have a matching coefficient, the one with the longer
+      //     vectorization length is used (sending equality to input order).
+      //   - Failing that, since neither of the inputs being compared are
+      //     going to succesfully advance the vectorizaion, we fall back to
+      //     order in the input map.
+      // - In all otther cases, ties are broken by position in the input map.
       auto &&zip = llvm::zip(params, upperDims);
       SmallVector<std::tuple<int64_t, uint32_t>> data(zip.begin(), zip.end());
-      std::sort(data.begin(), data.end());
+      std::sort(data.begin(), data.end(), [&](const auto &a, const auto &b) {
+        int64_t paramA = std::get<0>(a), paramB = std::get<0>(b);
+        uint32_t dimA = std::get<1>(a), dimB = std::get<1>(b);
+        if (paramA != paramB)
+          return paramA < paramB;
+        bool aHasVec = input[dimA].has_value(),
+             bHasVec = input[dimB].has_value();
+        if (!aHasVec && !bHasVec)
+          return dimA < dimB;
+        if (aHasVec && !bHasVec)
+          return true;
+        if (!aHasVec && bHasVec)
+          return false;
+
+        bool aIsGood = input[dimA]->needsCoefficient == paramA,
```

Comment:
nit : could use a better name here... as in I dont follow the relationship between being 'good' vs needed coefficient equating to embed coefficient.
Also it would be better use coeffA vs paramA to make things clearer...